### PR TITLE
[Merged by Bors] - chore(AlgebraicGeometry/EllipticCurve/Weierstrass): change the ext lemma for `EllipticCurve`

### PR DESCRIPTION
**Pull request by acmepjz**
Now it doesn't require that the discriminant is equal. This is like `Units.ext`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

discussion: https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2316828.20ext.20lemma.20for.20EllipticCurve/near/470380522

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean [hunk 1]
```diff
@@ -535,7 +534,29 @@ structure EllipticCurve (R : Type u) [CommRing R] extends WeierstrassCurve R whe
 
 namespace EllipticCurve
 
-variable {R : Type u} [CommRing R] (E : EllipticCurve R)
+variable {R : Type u} [CommRing R]
+
+section ext
+
+theorem ext' : Function.Injective (toWeierstrassCurve : EllipticCurve R → _)
```
Comment (Ruben-VandeVelde) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:541]: I suggest calling this `toWeierstrassCurve_injective`

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean [hunk 2]
```diff
@@ -535,7 +534,29 @@ structure EllipticCurve (R : Type u) [CommRing R] extends WeierstrassCurve R whe
...
+  | ⟨x1, _, x3⟩, ⟨y1, _, y3⟩, h => by
+    change x1 = y1 at h
+    congr
+    exact Units.ext (by rw [x3, y3, h])
+
+variable {x y : EllipticCurve R}
+
+theorem ext_iff' : x = y ↔ x.toWeierstrassCurve = y.toWeierstrassCurve :=
+  ⟨fun h ↦ by rw [h], fun h ↦ EllipticCurve.ext' h⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:550]: Then this is `toWeierstrassCurve_injective.eq_iff.symm`, if you need it. (Do you?)

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:550]: Good point! Actually I don't need it, though.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean [hunk 3]
```diff
@@ -535,7 +534,20 @@ structure EllipticCurve (R : Type u) [CommRing R] extends WeierstrassCurve R whe
...
+theorem toWeierstrassCurve_injective : Function.Injective (toWeierstrassCurve (R := R))
+  | ⟨x1, _, x3⟩, ⟨y1, _, y3⟩, h => by
+    change x1 = y1 at h
+    congr
+    exact Units.ext (by rw [x3, y3, h])
+
+@[ext]
+theorem ext {x y : EllipticCurve R} (h₁ : x.a₁ = y.a₁) (h₂ : x.a₂ = y.a₂) (h₃ : x.a₃ = y.a₃)
+    (h₄ : x.a₄ = y.a₄) (h₆ : x.a₆ = y.a₆) : x = y :=
+  toWeierstrassCurve_injective (WeierstrassCurve.ext h₁ h₂ h₃ h₄ h₆)
```
Comment (urkud) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:548]: Should it be this lemma or just `(h : x.toWeierstrassCurve = y.toWeierstrassCurve)`? In the latter case, you can tell `ext` to stop after this step and argue about bundled `WeierstrassCurve`s.

**Comment by urkud** [comment]
bors d=@Ruben-VandeVelde

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean (hunk 3 above)
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:548]: Currently the `ext` is this one together with `x.Δ' = y.Δ'` (which is not necessary and is removed in this PR). Yours is named as `ext'` (later renamed to `toWeierstrassCurve_injective`) in this PR. I don't know which one is better for the ext lemma, though.

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:548]: I've checked that
```lean
@[ext]
theorem ext : Function.Injective (toWeierstrassCurve (R := R))
```
works and the later proof using `ext` is not affected. Is this a good idea?

Comment (jcommelin) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:548]: I think the current version is better.

Comment (jcommelin) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:548]: As in: I think the PR is LGTM as it is.

**Comment by Ruben-VandeVelde** [comment]
I think this is fine, yeah.

bors merge

**Comment by acmepjz** [comment]
I think you should type `bors r+`

**Pull request closed without merging**